OpenRA.Mods.Common/Widgets/WidgetUtils.cs

```diff
@@ -270,6 +271,24 @@ public static void TruncateButtonToTooltip(ButtonWidget button, string text)
 			else
 				button.GetTooltipText = null;
 		}
+
+		public static void AddSuffixToPlayerNameLabel(LabelWidget label, Player p)
+		{
+			var client = p.World.LobbyInfo.ClientWithIndex(p.ClientIndex);
+			var nameFont = Game.Renderer.Fonts[label.Font];
+			var suffixLength = new CachedTransform<string, int>(s => nameFont.Measure(s).X);
+			var name = new CachedTransform<Pair<string, string>, string>(c =>
+				WidgetUtils.TruncateText(c.First, label.Bounds.Width - suffixLength.Update(c.Second), nameFont) + c.Second);
+
+			label.GetText = () =>
+			{
+				var suffix = p.WinState == WinState.Undefined ? "" : " (" + p.WinState + ")";
```

Comment:
This would be a good opportunity to save some memory churn (temporary strings if `p.WinState != WinState.Undefined`) by passing `p.WinState` into the `CachedTransform` and calculating the suffix only when it changes.

Reply:
Something like this:
```csharp
public static void BindPlayerNameAndStatus(LabelWidget label, Player p)
{
	var client = p.World.LobbyInfo.ClientWithIndex(p.ClientIndex);
	var nameFont = Game.Renderer.Fonts[label.Font];
	var suffixLength = new CachedTransform<string, int>(s => nameFont.Measure(s).X);
	var name = new CachedTransform<Tuple<string, WinState, Session.ClientState>, string>(c =>
	{
		var suffix = c.Item2 == WinState.Undefined ? "" : " (" + c.Item2 + ")";
		if (c.Item3 == Session.ClientState.Disconnected)
			suffix = " (Gone)";

		return TruncateText(c.Item1, label.Bounds.Width - suffixLength.Update(suffix), nameFont) + suffix;
	});

	label.GetText = () =>
	{
		var clientState = client != null ? client.State : Session.ClientState.Ready;
		return name.Update(Tuple.Create(p.PlayerName, p.WinState, clientState));
	};
}
```